sql/core/src/main/scala/org/apache/spark/sql/SparkSessionExtensions.scala

```diff
@@ -199,6 +199,21 @@ class SparkSessionExtensions {
     optimizerRules += builder
   }
 
+  private[this] val dataSourceRewriteRules = mutable.Buffer.empty[RuleBuilder]
+
+  private[sql] def buildDataSourceRewriteRules(session: SparkSession): Seq[Rule[LogicalPlan]] = {
+    dataSourceRewriteRules.map(_.apply(session))
```

Comment:
Shall we make it early `map` ? It can be potentially exposed to SPARK-16694 if we happened to lazily map in the future. We can either use a concrete return type or call like `.toArray`

e.g.)

```scala
scala> Seq(1, 2).view.map(_ => println("a")).isInstanceOf[Seq[_]]
res0: Boolean = true

scala> Seq(1, 2).map(_ => println("a")).isInstanceOf[Seq[_]]
a
a
res1: Boolean = true
```

Reply:
`dataSourceRewriteRules.map(_.apply(session))` need to call 'toSeq' to compatible with Scala 2.13

Reply:
@HyukjinKwon, I believe you are right this issue may happen if we switch from `mutable.Buffer` to something else in the future but I think we should update all places in `SparkSessionExtensions` if we want to fix it now. In this PR, I match the existing behavior for all other rules like `resolutionRuleBuilders` except calling `toSeq` like @LuciferYang pointed.

I'll add `toSeq` at the end to match all other places.

Reply:
We're running `Build and test / Scala 2.13 build with SBT (pull_request) ` in Github Actions build. As long as the build pass, we're all good. BTW, https://github.com/apache/spark/pull/30577#discussion_r534576769 causes a very annoying bug time to time. But sure, okay to leave it as is and fix all later.

Reply:
Added `toSeq`, let's see if it fixes the build.